File: python/p0004.py
from typing import List


class Solution:

  def findMedianSortedArrays(self, nums1: List[int],
                             nums2: List[int]) -> float:

    # Calculate the median indices
    median_indices = []
    total_length = len(nums1) + len(nums2)
    if total_length % 2 == 1:
        # If the total length is odd, there's only one median index
        median_indices = [total_length // 2]
    else:
        # If the total length is even, there are two median indices
        median_indices = [total_length // 2 - 1, total_length // 2]
    
    # Initialize pointers and current index
    index1 = 0
    index2 = 0
    current_index = -1
    current_median = None
    
    # Traverse both arrays until reaching the first median index
    while current_index < median_indices[0]:
        if index1 < len(nums1) and index2 < len(nums2):
            if nums1[index1] <= nums2[index2]:
                current_median = nums1[index1]
                index1 += 1
            else:
                current_median = nums2[index2]
                index2 += 1
        elif index1 < len(nums1):
            current_median = nums1[index1]
            index1 += 1
        else:
            current_median = nums2[index2]
            index2 += 1
        current_index += 1
    
    # If there's only one median index, return the current median
    if len(median_indices) == 1:
        return current_median
    else:
        # If there are two median indices, find the next median value
        second_median = None
        if index1 < len(nums1) and index2 < len(nums2):
            if nums1[index1] <= nums2[index2]:
                second_median = nums1[index1]
            else:
                second_median = nums2[index2]
        elif index1 < len(nums1):
            second_median = nums1[index1]
        else:
            second_median = nums2[index2]
        # Return the average of the two median values
        return (current_median + second_median) / 2

# Test case
solution = Solution()
print(solution.findMedianSortedArrays([1, 2], [3, 4]))

File: javascript/p0004.js
/**
 * @param {number[]} nums1
 * @param {number[]} nums2
 * @return {number}
 */
var findMedianSortedArrays = function (nums1, nums2) {
  // Calculate the median indices
  let medianIndices = [];
  if ((nums1.length + nums2.length) % 2) {
    // If the total length is odd, there's only one median index
    medianIndices = [Math.floor((nums1.length + nums2.length) / 2)];
  } else {
    // If the total length is even, there are two median indices
    medianIndices = [
      Math.floor((nums1.length + nums2.length) / 2) - 1,
      Math.floor((nums1.length + nums2.length) / 2),
    ];
  }

  // Initialize pointers and current index
  let index1 = 0;
  let index2 = 0;
  let currentIndex = -1;
  let currentMedian = null;

  // Traverse both arrays until reaching the first median index
  while (currentIndex < medianIndices[0]) {
    if (index1 < nums1.length && index2 < nums2.length) {
      if (nums1[index1] <= nums2[index2]) {
        currentMedian = nums1[index1];
        index1++;
      } else {
        currentMedian = nums2[index2];
        index2++;
      }
    } else if (index1 < nums1.length) {
      currentMedian = nums1[index1];
      index1++;
    } else {
      currentMedian = nums2[index2];
      index2++;
    }
    currentIndex++;
  }

  // If there's only one median index, return the current median
  if (medianIndices.length === 1) {
    return currentMedian;
  } else {
    // If there are two median indices, find the next median value
    let secondMedian = null;
    if (index1 < nums1.length && index2 < nums2.length) {
      if (nums1[index1] <= nums2[index2]) {
        secondMedian = nums1[index1];
      } else {
        secondMedian = nums2[index2];
      }
    } else if (index1 < nums1.length) {
      secondMedian = nums1[index1];
    } else {
      secondMedian = nums2[index2];
    }
    // Return the average of the two median values
    return (currentMedian + secondMedian) / 2;
  }
};

// Test case
console.log(findMedianSortedArrays([1, 2], [3, 4]));
